src/Helper/FasitHelper.php

```diff
@@ -0,0 +1,478 @@
+
+    [$certificateOptions, $tempCertFilename] = $this->getCertificateOptionsAndTempCertFilename();
+
+    // Attempt upload.
+    try {
+      $options = [
+        'headers' => [
+          'Content-Type' => 'multipart/form-data',
+          'X-Filename' => $originalFilename,
+          'X-Title' => pathinfo($originalFilename, PATHINFO_FILENAME),
+        ],
+        'multipart' => [
+          [
+            'name' => 'fileData',
+            'contents' => Utils::tryFopen($tempFilename, 'r'),
+            'filename' => pathinfo($tempFilename, PATHINFO_FILENAME),
+            'headers'  => [
+              'Content-Type' => '<Content-type header>',
+            ],
+          ],
+        ],
+        'cert' => $certificateOptions,
+      ];
+
+      $response = $this->client->request('POST', $endpoint, $options);
+    }
+    catch (GuzzleException $e) {
+      throw new FasitResponseException($e->getMessage(), $e->getCode());
+    } finally {
+      // Remove the certificate from disk.
+      if (file_exists($tempCertFilename)) {
+        unlink($tempCertFilename);
+      }
+      // Remove the attachment from disk.
+      if (file_exists($tempFilename)) {
+        unlink($tempFilename);
+      }
+    }
+
+    if (201 != $response->getStatusCode()) {
+      throw new FasitResponseException(sprintf('Expected status code 201, received %d', $response->getStatusCode()));
+    }
+
+    $content = json_decode($response->getBody()->getContents(), TRUE);
+    if (!isset($content['id'])) {
+      throw new FasitResponseException('Could not get upload id from response');
+    }
+
+    return ['filename' => $originalFilename, 'id' => $content['id']];
+  }
+
+  /**
+   * Uploads files from file elements to Fasit.
+   *
+   * @param string $submissionId
+   *   The submission id.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   * @throws \Drupal\os2forms_fasit\Exception\CertificateLocatorException
+   * @throws \Drupal\os2forms_fasit\Exception\FasitResponseException
+   * @throws \Drupal\os2forms_fasit\Exception\FileTypeException
+   *
+   * @phpstan-return array<mixed, mixed>
+   */
+  private function uploadFileElements(string $submissionId): array {
+    // Fetch element ids that may contain pdf files.
+    /** @var \Drupal\webform\Entity\WebformSubmission $submission */
+    $submission = $this->getSubmission($submissionId);
+    $fileIds = $this->getFileElementKeysFromSubmission($submission);
+    $fileStorage = $this->entityTypeManager->getStorage('file');
+
+    $uploads = [];
+
+    foreach ($fileIds as $fileId) {
+      /** @var \Drupal\file\Entity\File $file */
+      $file = $fileStorage->load($fileId);
+
+      // Ensure it is a pdf.
+      if (!empty(file_validate_extensions($file, 'pdf'))) {
+        throw new FileTypeException('Invalid file type uploaded. Only allowed file type is: pdf');
+      }
```

Comment:
This does not ensure that the file is a PDF, but only that its name ends with `.pdf`. You should use `$file->getMimeType()` to checks the file type.

Reply:
Updated.